File: test_dir/python/src/excpetions/throw_.py
def throw_fn():
    # raising an exception and catching all
    try:
        raise Exception("test")
    except Exception:
        print("exception happened")
    # raise an exception and catch specific
    try:
        raise TypeError("test")
    except TypeError:
        print("type error caught")
    # raise an exception and catch specific with the catch value
    try:
        raise TypeError("test")
    except TypeError as e:
        print("type error caught: " + str(e))
    # catch multiple types
    try:
        raise TypeError("test")
    except TypeError:
        print("type error caught")
    except ValueError:
        print("value error caught")
    except Exception:
        print("other error caught")

File: test_dir/cpp/src/excpetions/throw_.cpp
#include "excpetions\throw_.h"

void throw_fn() {
    try {
        throw PyppException(PyStr("test").str());
    } catch (const PyppException &) {
        PyStr("exception happened").print();
    }
    try {
        throw PyppTypeError(PyStr("test").str());
    } catch (const PyppTypeError &) {
        PyStr("type error caught").print();
    }
    try {
        throw PyppTypeError(PyStr("test").str());
    } catch (const PyppTypeError &pypp_e) {
        std::string e = pypp_e.what();
        (PyStr("type error caught: ") + to_pystr(e)).print();
    }
    try {
        throw PyppTypeError(PyStr("test").str());
    } catch (const PyppTypeError &) {
        PyStr("type error caught").print();
    } catch (const PyppValueError &) {
        PyStr("value error caught").print();
    } catch (const PyppException &) {
        PyStr("other error caught").print();
    }
}